File: BattleshipsinaBoard.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int countBattleships(vector<vector<char> >& board) {
        int count = 0;
        for(int i = 0; i < board.size(); i++) {
        	for(int j = 0; j < board[i].size(); j++) {
//        		if(board[i][j] == 'X' && !((j > 0 && board[i][j - 1] == 'X') || (i > 0 && board[i - 1][j] == 'X')))
        		if(board[i][j] == 'X' && ((j == 0 || board[i][j - 1] != 'X') && (i == 0 || board[i - 1][j] != 'X')))
        			count++;
			}
		}
		
		return count;
    }
};

int main() {
	
	int m, n;
	cin>>m>>n;
	
	vector<vector<char> > board;
	for(int i = 0; i < m; i++) {
		vector<char> row;
		for(int j = 0; j < n; j++) {
			char x;
			cin>>x;
			row.push_back(x);
		}
		board.push_back(row);
	}
	
	Solution *solution = new Solution();
	cout<<solution->countBattleships(board)<<endl;
	
	return 0;
}

File: BattleshipsinaBoard.py
class Solution:
    def countBattleships(self, board):
        """
        :type board: List[List[str]]
        :rtype: int
        """
        count = 0
        for i in range(len(board)):
            for j in range(len(board[i])):
                # if board[i][j] == 'X' and (i == 0 or board[i - 1][j] != 'X') and (j == 0 or board[i][j - 1] != 'X'):
                if board[i][j] == 'X' and not ((i > 0 and board[i - 1][j] == 'X') or (j > 0 and board[i][j - 1] == 'X')):
                    count += 1
        return count


if __name__ == '__main__':
    m = int(input())
    board = []
    for _ in range(m):
        board.append(list(input()))
    solution = Solution()
    print(solution.countBattleships(board))
